### PR TITLE
Fixed 29059 -- Made ChoiceWidget.optgroups() group non-grouped choices in null group.

### PR DESCRIPTION
**Pull request by gautamabhishek46**
https://code.djangoproject.com/ticket/29059

### CODE REVIEW

**Comment by timgraham** [comment]
Did you notice the failing tests? A new test that demonstrates what's fixed is also required.

**Comment by gautamabhishek46** [comment]
Should I change existing tests to suit my changes or write them as new?

**Comment by timgraham** [comment]
I haven't studied the changes to know if they're correct or not.

**Comment by carltongibson** [comment]
Hi @gautamabhishek46. 

Please have a look at the guidelines on [Submitting Patches](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/). There's a lot to help you there. 

To begin, here's what I notice instantly: 

* You need to rename the issue. Something like "Fixed #.... -- Description..." — please make it descriptive. 
* Then you need to make sure the existing tests pass. You need to look at the failures to see what's broken and how to fix that. 
* Then your new behaviour needs a new test case reproducing the original issue, that passes with your fix. 

See the [Patch Review Checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) for more. 

If you can resolve these issues it makes it possible for us to sensibly look at your patch. 

Thanks!

**Comment by gautamabhishek46** [comment]
Hi @carltongibson 

First of all, thank you for your reply.

Now all the existing test cases have passed, but I am not sure if new tests are required for my changes or not. If new tests are required then what tests I should write for my changes and how to write them. If you can provide some help on this then it will be a great help.

**Comment by timgraham** [comment]
A test is needed to show what behavior has changed as a result of the fix. I would expect the code to be somewhat similar to the steps to reproduce the issue on the ticket.

**Comment by carltongibson** [comment]
See the topic guide on [Writing and Running Tests](https://docs.djangoproject.com/en/2.0/topics/testing/overview/). It should give you what you need.

There are links there to tutorials on testing and lots of other advice.

**Comment by gautamabhishek46** [comment]
@carltongibson @timgraham Thanks for the reply again. Your replies helped me a lot.

I have added a test for my changes. Please take a look at it and let me know if any changes are required.

**Comment by gautamabhishek46** [comment]
Hi !!
Is there something wrong with my commits because it's been 3 days since I made my last commit for this issue. Please let me know if I need to change something.

**Comment by carltongibson** [comment]
Hi @gautamabhishek46 — Sorry for the delay: at any one point there are a significant number of patches to review, so it can take a while. You need to be patient I'm afraid. (I'm having a look now.) 

As per my first comment above, I've now edited the title of the issue to the `Fixed ... -- ...` format. This allows the PR to be picked up in the issue tracker (the issue now links to here) and so helps us to see that it's ready for review. 

Note: I'm not entirely convinced by the new issue title, but it'll do for a start.

File: tests/forms_tests/widget_tests/test_select.py [hunk 1]
```diff
@@ -354,6 +354,86 @@ def test_optgroups_integer_choices(self):
         label, options, index = groups[0]
         self.assertEqual(options[0]['value'], 0)
 
+    def test_optgroups_grouping_of_nogroup_choices(self):
```
Comment (carltongibson) [inline tests/forms_tests/widget_tests/test_select.py]: Is there any reason we can't just add an extra case to the existing `test_optgroups` method above, and so exercise this behaviour, without needing to add this extra method (which is largely identical)?

File: django/forms/widgets.py [hunk 2]
```diff
@@ -588,11 +589,15 @@ def optgroups(self, name, value, attrs=None):
                 group_name = option_value
                 subindex = 0
                 choices = option_label
+                groups.append((group_name, subgroup, index))
             else:
-                group_name = None
+                if subgroup_None is None:
+                    subgroup_None = []
+                    groups.append((None, subgroup_None, index))
```
Comment (carltongibson) [inline django/forms/widgets.py]: This is going to change the ordering of my choices. 

Take the example from the test: 

```
 + choices = [
 +            ("one", 1),
 +            [
 +                "italian", (
 +                    ("uno", 1),
 +                    ("due", 2)
 +                )
 +            ],
 +            ("two", 2),
 +        ]
```

I may deliberately want to present the choices in that order. An/some ungrouped default/common choices at the beginning, some groups, and then some ungrouped other choices at the end. 

I may have some grouped choices, then some ungrouped ones, then some more groups, and so on. With this change all the ungrouped choices would be in the second group, regardless of that ordering. 

I'm not sure we just want to enforce them all being grouped at whatever index the first ungrouped choice appears.

File: tests/forms_tests/widget_tests/test_select.py (hunk 1 above)
Comment (gautamabhishek46) [reply tests/forms_tests/widget_tests/test_select.py]: I also thought of adding an extra case to the existing test_optgroups method first. But when I searched on the internet, I found that many people are recommending to add a new test case. 

Maybe I misunderstood (adding a new test case == declaring a new function to reproduce the issue)

We can absolutely add another case in the test_optgroups method.

**Comment by timgraham** [comment]
That sounds fine to me. When I initially looked at the ticket, I wasn't sure a solution was feasible.

**Pull request closed without merging**